# Data Cleaning Mini Project: Ebay Car Sales Data

I will clean and analyze a dataset of used cars from *eBay Kleinanzeigen*, a classifieds section of the German eBay website. 

The dataset is from Kaggle and modified by Dataquest to add a more realistic dimension to the guided project. 

In [1]:
# import libraries
import numpy as np
import pandas as pd

In [3]:
# read csv file
autos = pd.read_csv('autos.csv', encoding = "Latin-1")

In [5]:
# take a look at the dataset
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

On the first look, we can make a few observations:

* some columns have null values, although not too many. 
* some columns have the wrong dtype (i.e. price is a string object, not a float).
* the column names use camelcase instead of snakecase.

We will clean the data to address these observations.

### Task: Switch from Camelcase to Snakecase

First, we will switch the column names from camelcase to snakecase. we first take a look at the column names, and replace them so that everything is in lowercase and that underscores are inserted at appropriate places.

In [7]:
# original column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [8]:
# new column names
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [9]:
# check results
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Task: Observation 

Next, we take a look at the dataset using df.describe() method.

In [10]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Some observations:

* sellers are almost always private
* offer type is almost always "Angebot"
* we don't have descriptive statistics for price because they are not in numeric type yet
* same goes with odometer.
* registration year is in float and seems to have weird numbers. For example, the smallest registration year is 1000, but cars did not exist in the year 1000! 
* number of pictures and postal code don't have values for unique, top, and freq. Weird. 

Let's start with changing the values in `price` and `odometer` from text to numeric values.
### Task: Text to Numeric

In [12]:
# remove any non-numeric characters
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")

# convert the columns to numeric dtype
autos["price"] = autos["price"].astype(float)
autos["odometer"] = autos["odometer"].astype(int)

# rename columns
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

Now, we'll look for any data in these two columns that doesn't look right.

In [14]:
autos["price"].shape

(50000,)

In [15]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
autos["price"].value_counts()

0.0           1421
500.0          781
1500.0         734
2500.0         643
1200.0         639
1000.0         639
600.0          531
800.0          498
3500.0         498
2000.0         460
999.0          434
750.0          433
900.0          420
650.0          419
850.0          410
700.0          395
4500.0         394
300.0          384
2200.0         382
950.0          379
1100.0         376
1300.0         371
3000.0         365
550.0          356
1800.0         355
5500.0         340
1250.0         335
350.0          335
1600.0         327
1999.0         322
              ... 
2225.0           1
69997.0          1
139997.0         1
69999.0          1
4780.0           1
8930.0           1
21599.0          1
15911.0          1
10000000.0       1
5180.0           1
919.0            1
1247.0           1
5998.0           1
27020.0          1
21888.0          1
46500.0          1
2001.0           1
2459.0           1
345000.0         1
34940.0          1
2785.0           1
5248.0      

In [17]:
autos["odometer_km"].shape

(50000,)

In [18]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

The odometer column looks fine. The price column, though, seems to contain some abnormal values. Since it is unlikely that a used car will cost more than $30,000 or be given out for free, let's remove these outliers. We can bring them back later if needed.

In [50]:
# remove price outliers
autos = autos[autos["price"].between(0,30000)]

In [22]:
# observe remaining values
autos["price"].shape

(49800,)

In [51]:
autos["price"].describe()

count    43677.000000
mean      5226.875197
std       5762.973688
min          0.000000
25%       1200.000000
50%       3000.000000
75%       7200.000000
max      30000.000000
Name: price, dtype: float64

That looks more reasonable.

### Task: Distribution of Dates

We calculate the distribution of values in `date_crawled`, `ad_created`, `last_seen` columns as percentages. We will make some observations

In [30]:
autos["date_crawled"].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05 14:06:30    0.00002
2016-03-05 14:06:40    0.00002
2016-03-05 14:07:04    0.00002
2016-03-05 14:07:08    0.00002
2016-03-05 14:07:21    0.00002
2016-03-05 14:07:26    0.00002
2016-03-05 14:07:40    0.00002
2016-03-05 14:07:45    0.00002
2016-03-05 14:08:00    0.00004
2016-03-05 14:08:05    0.00004
2016-03-05 14:08:27    0.00002
2016-03-05 14:08:42    0.00002
2016-03-05 14:09:02    0.00004
2016-03-05 14:09:05    0.00002
2016-03-05 14:09:20    0.00002
2016-03-05 14:09:22    0.00002
2016-03-05 14:09:38    0.00002
2016-03-05 14:09:46    0.00002
2016-03-05 14:09:56    0.00002
2016-03-05 14:09:57    0.00002
2016-03-05 14:09:58    0.00004
2016-03-05 14:10:18    0.00002
2016-03-05 14:10:20    0.00002
2016-03-05 14:10:46    0.00002
2016-03-05 14:11:03    0.00002
2016-03-05 14:11:05    0.00002
2016-03-05 14:11:14    0.00002
2016-03-05 14:11:15    0.00002
2016-03-05 14:11:25    0.00002
2016-03-05 14:11:40    0.00002
                        ...   
2016-04-07 10:36:19    0.00002
2016-04-

In [31]:
autos["ad_created"].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11 00:00:00    0.000020
2015-08-10 00:00:00    0.000020
2015-09-09 00:00:00    0.000020
2015-11-10 00:00:00    0.000020
2015-12-05 00:00:00    0.000020
2015-12-30 00:00:00    0.000020
2016-01-03 00:00:00    0.000020
2016-01-07 00:00:00    0.000020
2016-01-10 00:00:00    0.000040
2016-01-13 00:00:00    0.000020
2016-01-14 00:00:00    0.000020
2016-01-16 00:00:00    0.000020
2016-01-22 00:00:00    0.000020
2016-01-27 00:00:00    0.000060
2016-01-29 00:00:00    0.000020
2016-02-01 00:00:00    0.000020
2016-02-02 00:00:00    0.000040
2016-02-05 00:00:00    0.000040
2016-02-07 00:00:00    0.000020
2016-02-08 00:00:00    0.000020
2016-02-09 00:00:00    0.000040
2016-02-11 00:00:00    0.000020
2016-02-12 00:00:00    0.000060
2016-02-14 00:00:00    0.000040
2016-02-16 00:00:00    0.000020
2016-02-17 00:00:00    0.000020
2016-02-18 00:00:00    0.000040
2016-02-19 00:00:00    0.000060
2016-02-20 00:00:00    0.000040
2016-02-21 00:00:00    0.000060
                         ...   
2016-03-

In [32]:
autos["last_seen"].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05 14:45:46    0.00002
2016-03-05 14:46:02    0.00002
2016-03-05 14:49:34    0.00002
2016-03-05 15:16:11    0.00002
2016-03-05 15:16:47    0.00002
2016-03-05 15:28:10    0.00002
2016-03-05 15:41:30    0.00002
2016-03-05 15:45:43    0.00002
2016-03-05 15:47:38    0.00002
2016-03-05 15:47:44    0.00002
2016-03-05 16:45:57    0.00002
2016-03-05 16:47:28    0.00002
2016-03-05 17:15:45    0.00002
2016-03-05 17:16:12    0.00002
2016-03-05 17:16:14    0.00002
2016-03-05 17:16:23    0.00002
2016-03-05 17:17:02    0.00002
2016-03-05 17:39:19    0.00002
2016-03-05 17:40:14    0.00002
2016-03-05 17:44:50    0.00002
2016-03-05 17:44:54    0.00002
2016-03-05 17:46:01    0.00002
2016-03-05 18:17:58    0.00002
2016-03-05 18:47:14    0.00002
2016-03-05 18:50:38    0.00002
2016-03-05 19:15:08    0.00002
2016-03-05 19:15:20    0.00002
2016-03-05 19:15:42    0.00002
2016-03-05 19:16:36    0.00002
2016-03-05 19:17:17    0.00002
                        ...   
2016-04-07 14:58:09    0.00004
2016-04-

Observations: 

* the crawler has been making consistent contributions throughout the time period
* there are more ads in recent dates than in older dates
* the dates last seen are evenly distributed.

In [33]:
autos["registration_year"].describe()

count    49800.000000
mean      2005.074398
std        105.920807
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration years give statistics that are way off, ranging from 1000 to 9999. We need some cleaning to do. 

We let the lowest acceptable values of the registration year to be 1995, since that is the year that eBay was founded. We set the highest acceptable values of the resgiration to be 2016, since the most recent listing ends in 2016. 

In [34]:
# remove values that are out of acceptable year range
autos = autos[autos["registration_year"].between(1995,2016)]

# calculate the distribution of remaining values
autos["registration_year"].value_counts(normalize=True)

2000    0.075793
2005    0.068150
1999    0.067766
2004    0.061819
2003    0.061638
2006    0.061186
2001    0.061073
2002    0.057251
1998    0.055375
2007    0.051938
2008    0.050197
2009    0.047302
1997    0.045788
2011    0.036766
2010    0.035861
1996    0.032605
1995    0.029621
2016    0.029575
2012    0.029530
2013    0.017750
2014    0.014675
2015    0.008344
Name: registration_year, dtype: float64

Roughly speaking, it seems like there are more listings of older cars, especially registered in 2000, than there are of newer cars (2012~2016). 

### Task: Aggregating by Brand

In [52]:
# what brands are there and how many?
brands = autos["brand"].value_counts()
brands

volkswagen        9036
opel              4923
bmw               4777
mercedes_benz     3924
audi              3700
ford              3180
renault           2226
peugeot           1385
fiat              1178
seat               856
skoda              761
mazda              689
nissan             677
smart              667
citroen            640
toyota             567
hyundai            472
mini               408
volvo              386
mitsubishi         371
kia                341
honda              332
sonstige_autos     310
alfa_romeo         292
suzuki             263
chevrolet          220
chrysler           162
dacia              123
daihatsu           120
porsche            109
subaru              94
jeep                84
daewoo              72
land_rover          71
saab                68
rover               63
jaguar              61
lancia              45
lada                22
trabant              2
Name: brand, dtype: int64

In [53]:
# let's choose top 20 brands
top_20_brands = brands.head(20).index
top_20_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'hyundai', 'mini', 'volvo', 'mitsubishi'],
      dtype='object')

In [54]:
# compute aggregate means
means = {}
for b in top_20_brands:
    selected_rows = autos[autos["brand"] == b]
    b_mean = selected_rows["price"].mean()
    means[b] = b_mean  
means

{'audi': 8197.365945945947,
 'bmw': 7514.738329495499,
 'citroen': 3579.7640625,
 'fiat': 2656.483870967742,
 'ford': 3353.351257861635,
 'hyundai': 5319.576271186441,
 'mazda': 3910.2278664731493,
 'mercedes_benz': 7624.168450560653,
 'mini': 10419.622549019608,
 'mitsubishi': 3264.7924528301887,
 'nissan': 4772.227474150664,
 'opel': 2912.770668291692,
 'peugeot': 3082.9761732851985,
 'renault': 2368.7327044025155,
 'seat': 4328.5490654205605,
 'skoda': 6350.352168199737,
 'smart': 3545.544227886057,
 'toyota': 5077.0,
 'volkswagen': 5303.9055998229305,
 'volvo': 4935.860103626943}

There is a great variation in mean price depending on the brand of the car. For example, if a car is a renault, the average price is \$2368, whereas if it is a mini, then it is \$10419 on average.

In [64]:
# mean mileage and price for top brands
mileage_mean = {}
price_mean = {}
for b in top_20_brands:
    selected_rows = autos[autos["brand"] == b]
    mean_mileage = selected_rows["odometer_km"].mean()
    mean_price = selected_rows["price"].mean()
    mileage_mean[b] = mean_mileage
    price_mean[b] = mean_price

# construct series objects with the dictionaries
mileage_avg = pd.Series(mileage_mean)
price_avg = pd.Series(price_mean)

# create dataframe
avg_data = pd.DataFrame(mileage_avg)
avg_data["price_avg"] = price_avg

# pretty print
avg_data.columns = {"Average Mileage","Average Price"}
avg_data.sort_values("Average Mileage",axis=0,ascending=False,inplace=True)
avg_data


Average Mileage  Average Price
volvo            138523.316062    4935.860104
bmw              134122.880469    7514.738329
mercedes_benz    132454.128440    7624.168451
audi             131681.081081    8197.365946
opel             129971.562056    2912.770668
volkswagen       129019.477645    5303.905600
renault          128481.581312    2368.732704
mitsubishi       127291.105121    3264.792453
peugeot          127054.151625    3082.976173
ford             125520.440252    3353.351258
mazda            125232.220610    3910.227866
seat             122091.121495    4328.549065
citroen          120289.062500    3579.764062
fiat             118196.095076    2656.483871
nissan           117363.367799    4772.227474
toyota           115458.553792    5077.000000
skoda            111563.731932    6350.352168
hyundai          106419.491525    5319.576271
smart             99595.202399    3545.544228
mini              88970.588235   10419.622549

We can see that the brand with the lowest average mileage--mini--has the highest average price. However, this does not translate to a general trend of 'the lower the mileage, the higher the price.' 